src/Compilers/CSharp/Test/Emit2/Semantics/RefStructInterfacesTests.cs

```diff
@@ -5316,5 +5316,846 @@ void verify(ModuleSymbol m)
+        Test4((long)4);
+    }
+
+    static void Test1<T>(T x) where T : allows ref struct
+    {
+        System.Console.WriteLine(""Called {0}"", typeof(T));
+    }
+
+    static void Test2<T>(T x)
+    {
+        Test1(x);
+        Test1<T>(x);
+    }
+
+    static void Test3<T>(T x) where T : allows ref struct
+    {
+        Test1(x);
+        Test1<T>(x);
+    }
+
+    static void Test4<T>(T x)
+    {
+        Test2(x);
+        Test2<T>(x);
+    }
+}
+";
+            var comp = CreateCompilation(src, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics, options: TestOptions.ReleaseExe);
+
+            CompileAndVerify(comp, expectedOutput: !ExecutionConditionUtil.IsMonoOrCoreClr ? null : @"
+Called System.Byte
+Called System.Byte
+Called System.Int32
+Called System.Int32
+Called S1
+Called S1
+Called System.Int64
+Called System.Int64
+Called System.Int64
+Called System.Int64", verify: ExecutionConditionUtil.IsMonoOrCoreClr ? Verification.Passes : Verification.Skipped).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void ConstraintsCheck_05()
+        {
+            var src = @"
+ref struct S1
+{
+}
+
+class C<T, S>
+    where T : allows ref struct
+    where S : T
+{
+    static void Main()
+    {
+        _ = typeof(C<S1, S1>);
+        _ = typeof(C<int, int>);
+        _ = typeof(C<object, object>);
+        _ = typeof(C<object, string>);
+    }
+}
+";
+            var comp = CreateCompilation(src, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
+
+            comp.VerifyDiagnostics(
+                // (12,26): error CS9504: The type 'S1' may not be a ref struct or a type parameter allowing ref structs in order to use it as parameter 'S' in the generic type or method 'C<T, S>'
+                //         _ = typeof(C<S1, S1>);
+                Diagnostic(ErrorCode.ERR_NotRefStructConstraintNotSatisfied, "S1").WithArguments("C<T, S>", "S", "S1").WithLocation(12, 26)
+                );
+        }
+
+        [Fact]
+        public void IllegalBoxing_01()
+        {
+            var src = @"
+public class Helper
+{
+    public static object Test<T>(T x) where T : allows ref struct
+    {
+        return x;
+    }
+}
+";
+            var comp = CreateCompilation(src, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
+
+            // PROTOTYPE(RefStructInterfaces): The boxing should be disallowed.
+            comp.VerifyDiagnostics(
+                );
+        }
+
+        [Fact]
+        public void IllegalBoxing_02()
+        {
+            var src = @"
+public interface I1
+{
+}
+
+public class Helper
+{
+    public static I1 Test<T>(T x) where T : I1, allows ref struct
+    {
+        return x;
+    }
+}
+";
+            var comp = CreateCompilation(src, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
+
+            // PROTOTYPE(RefStructInterfaces): The boxing should be disallowed.
+            comp.VerifyDiagnostics(
+                );
+        }
```

Comment:
>} [](http://example.com/codeflow?start=8&length=1)

Consider testing explicit cast as well: `return (I1)x;`.

Also:
```csharp
static U Test1<T, U>(T x)
    where T : allows ref struct
    where U : T
{
    return x;
}
static U Test2<T, U>(T x)
    where T : allows ref struct
    where U : T
{
    return (U)x;
}
```

Reply:
> Consider testing explicit cast as well: return (I1)x;.

Testing boxing is not a goal of this PR. It is known that proper checks are not in place and the added tests a sufficient to make the knowledge obvious